validator-network/src/network_impl.rs

```diff
@@ -76,50 +88,111 @@ where
-                .clone();
+    ) -> Result<N::PeerId, NetworkError<N::Error>>
+    where
+        <N as Network>::Error: Sync + Send,
+    {
+        let public_key = self
+            .validator_keys
+            .read()
+            .get(usize::from(validator_id))
+            .ok_or(NetworkError::UnknownValidator(validator_id))?
+            .clone();
 
-            let peer_id = self
+        let update_peer_cache = {
+            if let Some(cache_state) = self
                 .validator_peer_id_cache
                 .read()
                 .get(&public_key.compressed().clone())
-                .cloned();
-            (peer_id, public_key)
+            {
+                match cache_state {
+                    CacheState::Resolved(peer_id) => return Ok(*peer_id),
+                    CacheState::Error(_) => true,
+                    CacheState::InProgress => {
+                        log::debug!(validator_id, "Record resolution is in progress");
+                        false
+                    }
+                }
+            } else {
+                log::error!(
+                    ?public_key,
+                    validator_id,
+                    "Could not find peer ID for validator in DHT",
+                );
+                true
+            }
         };
-
-        if let Some(peer_id) = peer_id {
-            Ok(peer_id)
-        } else if let Some(peer_id) = Self::resolve_peer_id(&self.network, &public_key).await? {
-            self.validator_peer_id_cache
-                .write()
-                .insert(public_key.compressed().clone(), peer_id);
-            Ok(peer_id)
-        } else {
-            log::error!(
-                "Could not find peer ID for validator in DHT: public_key = {:?}",
-                public_key
-            );
-            Err(NetworkError::UnknownValidator(validator_id))
+        if update_peer_cache {
+            // Cache is empty for this validator ID, query the entry
+            {
+                self.validator_peer_id_cache
+                    .write()
+                    .insert(public_key.compressed().clone(), CacheState::InProgress);
+            }
+            let self_ = self.arc_clone();
+            tokio::spawn(async move {
+                Self::update_peer_id_cache(&self_, validator_id, &public_key).await;
+            });
         }
+        Err(NetworkError::UnknownValidator(validator_id))
     }
 
-    /// Clears the validator->peer_id cache.
+    /// Clears the validator->peer_id cache on a `RequestError`.
     /// The cached entry should be cleared when the peer id might have changed.
-    fn clear_validator_peer_id_cache(&self, validator_id: u16) {
-        if let Some(validator_key) = self.validator_keys.read().get(usize::from(validator_id)) {
-            self.validator_peer_id_cache
-                .write()
-                .remove(validator_key.compressed());
+    fn clear_validator_peer_id_cache_on_error(&self, validator_id: u16, error: RequestError) {
```

Comment:
```suggestion
    fn clear_validator_peer_id_cache_on_error(&self, validator_id: u16, error: &RequestError) {
```
reduces necessary cloning.